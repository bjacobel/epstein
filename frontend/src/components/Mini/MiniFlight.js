import React from 'react';
import { useQuery, gql } from '@apollo/client';
import { Link } from 'react-router-dom';
import { format, parseISO } from 'date-fns';

import Loading from '../Loading';
import ErrorBoundary from '../ErrorBoundary';
import logError from '../../utils/errors';
import { box, link, row } from './mini.css';
import { robotoMono as mono } from '../../stylesheets/shared.css';

const FLIGHT = gql`
  query flight($id: Int!) {
    flight(id: $id) {
      id
      date
      source {
        iata_code
      }
      destination {
        iata_code
      }
      passengers {
        pageInfo {
          count
        }
      }
    }
  }
`;

export default ({ id, done }) => {
  const { loading, error, data } = useQuery(FLIGHT, {
    variables: { id },
    onCompleted: done,
    onError: done,
  });

  if (loading) return done ? null : <Loading />;
  if (error) {
    logError(error);
    return null;
  }

  return (
    <ErrorBoundary>
      <div className={box}>
        <Link to={`/flight/${data.flight.id}`} className={link}>
          <div className={row}>
            <span>{format(parseISO(data.flight.date), 'MMM d, y')}</span>
            <span>{`${data.flight.passengers.pageInfo.count} passengers`}</span>
          </div>
          <div className={row}>
            <span>
              <span className={mono}>{data.flight.source.iata_code}</span>
              <span> to </span>
              <span className={mono}>{data.flight.destination.iata_code}</span>
            </span>
          </div>
        </Link>
      </div>
    </ErrorBoundary>
  );
};
