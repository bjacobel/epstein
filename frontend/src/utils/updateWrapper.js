import { act } from 'react-dom/test-utils';

export function wait(amount = 0) {
  return new Promise(resolve => setTimeout(resolve, amount));
}

export default async (wrapper, amount = 0) => {
  await act(async () => {
    jest.useRealTimers();
    await wait(amount);
    jest.useFakeTimers();
    wrapper.update();
  });
};
