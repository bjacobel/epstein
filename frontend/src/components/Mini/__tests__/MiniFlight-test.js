import React from 'react';
import { ApolloProvider, ApolloClient, InMemoryCache } from '@apollo/client';
import { mount } from 'enzyme';

import MockLink from '../../../utils/testing/MockLink';
import updateWrapper from '../../../utils/testing/updateWrapper';
import MiniFlight, { FLIGHT } from '../MiniFlight';

const data = {
  flight: {
    id: 206,
    date: '2003-04-06',
    source: {
      iata_code: 'GRU',
    },
    destination: {
      iata_code: 'LBG',
    },
    passengers: {
      pageInfo: {
        count: 5,
      },
    },
  },
};

const fullPassengers = {
  edges: [
    {
      name: 'Jeffrey Epstein',
      slug: 'jeffrey-epstein',
      __typename: 'VerifiedPassenger',
    },
    {
      name: 'Ghislane Maxwell',
      slug: 'ghislane-maxwell',
      __typename: 'VerifiedPassenger',
    },
    {
      name: 'Sean Koo',
      slug: 'sean-koo',
      __typename: 'VerifiedPassenger',
    },
    {
      name: 'Jean-Luc Brunel',
      slug: 'jean-luc-brunel',
      __typename: 'VerifiedPassenger',
    },
    {
      literal: 'MAGALE BLACHOU',
      __typename: 'LiteralPassenger',
    },
  ],
};

const errors = [
  {
    path: ['flight', 'destination'],
    data: {
      iata_code: null,
    },
    errorType: 'NotFound',
    errorInfo: {
      query: 'GVAL',
    },
    locations: [
      {
        line: 8,
        column: 5,
        sourceName: null,
      },
    ],
    message: "Couldn't find Airfield with IATA, GPS or ident code GVAL",
  },
];

describe('MiniFlight viewer', () => {
  let client, link;

  beforeEach(() => {
    link = new MockLink();
    client = new ApolloClient({
      cache: new InMemoryCache({
        addTypename: false,
      }),
      link,
    });
  });

  describe('standard mode', () => {
    beforeEach(() => {
      const queryHandler = jest.fn(() => Promise.resolve({ data }));
      link.setRequestHandler(FLIGHT, queryHandler);
    });

    it('matches snapshot', async () => {
      const wrapper = mount(
        <ApolloProvider client={client}>
          <MiniFlight id={205} />
        </ApolloProvider>,
      );
      await updateWrapper(wrapper);
      expect(wrapper.find(MiniFlight)).toMatchSnapshot();
    });
  });

  describe('search results mode', () => {
    beforeEach(() => {
      const enrichedData = { ...data };
      Object.assign(enrichedData.flight.passengers, fullPassengers);
      const queryHandler = jest.fn(() => Promise.resolve({ data: enrichedData }));
      link.setRequestHandler(FLIGHT, queryHandler);
    });

    it('matches snapshot', async () => {
      const wrapper = mount(
        <ApolloProvider client={client}>
          <MiniFlight id={205} fullManifest />
        </ApolloProvider>,
      );
      await updateWrapper(wrapper);
      expect(wrapper.find(MiniFlight)).toMatchSnapshot();
    });
  });

  describe('when destination airport code is not found', () => {
    beforeEach(() => {
      const errData = { ...data };
      errData.flight.destination = null;
      const queryHandler = jest.fn(() => Promise.resolve({ data: errData, errors }));
      link.setRequestHandler(FLIGHT, queryHandler);
    });

    it('renders fallback for that location', async () => {
      const wrapper = mount(
        <ApolloProvider client={client}>
          <MiniFlight id={205} />
        </ApolloProvider>,
      );
      await updateWrapper(wrapper);
      expect(
        wrapper
          .find('.row')
          .at(1)
          .text(),
      ).toEqual('GRU to GVAL');
    });
  });
});
