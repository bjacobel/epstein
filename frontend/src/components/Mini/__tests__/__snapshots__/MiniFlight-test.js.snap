// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiniFlight viewer search results mode matches snapshot 1`] = `
<_default
  fullManifest={true}
  id={205}
>
  <ErrorBoundary>
    <div
      className="box"
    >
      <Link
        className="link"
        to="/flight/206"
      >
        <a
          href="/flight/206"
        >
          <div
            className="row"
          >
            <span>
              Apr 6, 2003
            </span>
            <span>
              <span
                className="robotoMono"
              >
                GRU
              </span>
              <span>
                 to 
              </span>
              <span
                className="robotoMono"
              >
                LBG
              </span>
            </span>
          </div>
          <div
            className="row"
          >
            <div
              className="leftJustifyRow"
            >
              <span
                key="jeffrey-epstein"
              >
                Jeffrey Epstein, 
              </span>
              <span
                key="ghislane-maxwell"
              >
                Ghislane Maxwell, 
              </span>
              <span
                key="sean-koo"
              >
                Sean Koo, 
              </span>
              <span
                key="jean-luc-brunel"
              >
                Jean-Luc Brunel, 
              </span>
              <span
                className="greyName"
                key="MAGALE BLACHOU"
              >
                MAGALE BLACHOU
              </span>
            </div>
          </div>
        </a>
      </Link>
    </div>
  </ErrorBoundary>
</_default>
`;

exports[`MiniFlight viewer standard mode matches snapshot 1`] = `
<_default
  id={205}
>
  <ErrorBoundary>
    <div
      className="box"
    >
      <Link
        className="link"
        to="/flight/206"
      >
        <a
          href="/flight/206"
        >
          <div
            className="row"
          >
            <span>
              Apr 6, 2003
            </span>
            <span>
              5 passengers
            </span>
          </div>
          <div
            className="row"
          >
            <span>
              <span
                className="robotoMono"
              >
                GRU
              </span>
              <span>
                 to 
              </span>
              <span
                className="robotoMono"
              >
                LBG
              </span>
            </span>
          </div>
        </a>
      </Link>
    </div>
  </ErrorBoundary>
</_default>
`;
